gateway-util-launcher/src/main/java/org/apache/knox/gateway/launcher/Extender.java

```diff
@@ -0,0 +1,77 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.knox.gateway.launcher;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Properties;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Extender {
+
+    private static final String CLASSPATH_EXTENSION_PROPERTY = "gateway.server.classpath.extension";
+    private static final String CLASSPATH_PROPERTY_PATTERN = "<property>\\s*<name>" + CLASSPATH_EXTENSION_PROPERTY + "</name>\\s*<value>(.*?)</value>\\s*</property>";
+    private static final String CONFIG_FILE = "gateway-site.xml";
+    private static final String CONFIG_PATH = "../conf/" + CONFIG_FILE;
+    private static final String CLASS_PATH_PROPERTY = "class.path";
+    private static final String MAIN_CLASS_PROPERTY = "main.class";
+    private static final String GATEWAY_SERVER_MAIN_CLASS = "org.apache.knox.gateway.GatewayServer";
+    private static final String[] CLASS_PATH_DELIMITERS = new String[]{",", ";"};
+
+    private final File base;
+    private final Properties properties;
+
+    public Extender(File base, Properties properties) {
+        this.base = base;
+        this.properties = properties;
+    }
+
+    public void extendClassPathProperty() throws IOException {
+        Path configFilePath = Paths.get(base.getPath(), CONFIG_PATH);
+        if (GATEWAY_SERVER_MAIN_CLASS.equals(properties.getProperty(MAIN_CLASS_PROPERTY)) && Files.isReadable(configFilePath)) {
+            String configContent = new String(Files.readAllBytes(configFilePath), StandardCharsets.UTF_8);
+            extractExtensionPathIntoProperty(configContent);
+        }
+    }
+
+    protected void extractExtensionPathIntoProperty(String configContent) {
+        Pattern pattern = Pattern.compile(CLASSPATH_PROPERTY_PATTERN, Pattern.DOTALL);
+        Matcher matcher = pattern.matcher(configContent);
+
+        if (matcher.find()) {
+            StringBuilder newClassPath = new StringBuilder(matcher.group(1).trim());
+            if (newClassPath.length() > 0) {
+                if (!endsWithDelimiter(newClassPath.toString())) {
+                    newClassPath.append(CLASS_PATH_DELIMITERS[1]);
+                }
+                newClassPath.append(properties.getProperty(CLASS_PATH_PROPERTY));
+                properties.setProperty(CLASS_PATH_PROPERTY, newClassPath.toString());
+            }
```

Comment:
Maybe adding a WARN log message here. So that users who explicitly configured something will not be surprised their configuration doesn't take effect.

Reply:
I agree it would be nice to have some kind of feedback for the user. However there is no any kind of logging setup in the launcher, there is no log dependencies added either to keep it lightweight. At this point the log4j properties are not loaded. Only the standard output is available.

Reply:
Ack.